### PR TITLE
Introduce class for setting up debug integration

### PR DESCRIPTION
**Pull request by nikolay-egorov**
This way, newer version of kernel will keep thread for creating breakpoints

### CODE REVIEW

File: jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/creating/LazilyConstructibleReplComponentsProvider.kt [hunk 1]
```diff
@@ -29,8 +30,10 @@ abstract class LazilyConstructibleReplComponentsProvider : ReplComponentsProvide
     override val homeDir: File? by lazy { provideHomeDir() }
     protected abstract fun provideHomeDir(): File?
 
-    override val debugPort: Int? by lazy { provideDebugPort() }
-    protected abstract fun provideDebugPort(): Int?
+    override val debugPort: Int? by lazy { debugIntegrationProvider.provideDebugPort() }
```
Comment (ileasile) [inline jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/creating/LazilyConstructibleReplComponentsProvider.kt:33]: No, that's not how this pattern works. You can remove `val debugPort` completely if you wish. But just observe the code: is there any lazy implementation other than `lazy { provide*() }`?

File: jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/debug/DebugIntegrationProviderMock.kt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package org.jetbrains.kotlinx.jupyter.debug
+
+object DebugIntegrationProviderMock : DebugIntegrationProvider(null)
```
Comment (ileasile) [inline jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/debug/DebugIntegrationProviderMock.kt:3]: Looks more like a singleton property bc you actually don't override anything

File: jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/debug/DebugIntegration.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    }
+
+    fun provideDebugPort(): Int? = debugPort
+
+    fun provideSuspendedThread(runnable: () -> Unit): Thread {
+        return thread(start = provideDebugPort() != null, name = DEBUG_THREAD_NAME, block = runnable)
+    }
+}
+
+class NotebookDebugIntegrationProvider(
```
Comment (ileasile) [inline jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/debug/DebugIntegration.kt:19]: If it's a `NotebookDebugIntegrationProvider` then why `DebugIntegrationProvider` is missing "Notebook" part in its name?

File: src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/creating/creatingUtil.kt [hunk 4]
```diff
@@ -37,7 +39,9 @@ fun createRepl(
         override fun provideIsEmbedded() = isEmbedded
         override fun provideDisplayHandler() = displayHandler
         override fun provideCommunicationFacility(): JupyterCommunicationFacility = communicationFacility
-        override fun provideDebugPort(): Int? = debugPort
+        override fun provideDebugIntegration(): DebugIntegrationProvider {
+            return NotebookDebugIntegrationProvider(debugPort)
```
Comment (ileasile) [inline src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/creating/creatingUtil.kt:43]: Once again, is there anything around this code that doesn't simply use passed constructor parameter?

File: jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/debug/DebugIntegration.kt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import kotlin.concurrent.thread
+
+abstract class DebugIntegrationProvider(
+    protected val debugPort: Int?,
+) {
+    companion object {
+        const val DEBUG_THREAD_NAME = "NotebookDebugThread"
+    }
+
+    fun provideDebugPort(): Int? = debugPort
```
Comment (ileasile) [inline jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/debug/DebugIntegration.kt:12]: It looks like debugPort doesn't need wrappers like this

Comment (ileasile) [reply jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/debug/DebugIntegration.kt:12]: Or maybe we don't need to provide debug port here at all? It looks like the logic is boolean here: if we want or don't want to create a debug thread

File: jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/debug/DebugIntegration.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+    fun provideSuspendedThread(runnable: () -> Unit): Thread {
```
Comment (ileasile) [inline jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/debug/DebugIntegration.kt:14]: This abstraction looks leaking. I would suggest to have just a boolean flag here and create the thread in place it's needed

**Review by ileasile — CHANGES_REQUESTED** [review]
For now it looks like we don't need debugPort in options at all. We need boolean instead. Abstraction DebugIntegrationProvider seems not needed

File: jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/creating/LazilyConstructibleReplComponentsProvider.kt (hunk 1 above)
Comment (nikolay-egorov) [reply jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/creating/LazilyConstructibleReplComponentsProvider.kt:33]: I agree, that's a violation of the pattern. Probably, we might need debug port for future cases in embedded kernel

File: src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/creating/creatingUtil.kt (hunk 4 above)
Comment (nikolay-egorov) [reply src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/creating/creatingUtil.kt:43]: True, can be done. I'll change. However, I didn't think about this as a pattern to follow

**Comment by nikolay-egorov** [comment]
I'll think about it one more time. Now it indeed seems like we don't need abstraction for debug just yet

File: jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/messaging/MessageHandler.kt [hunk 7]
```diff
@@ -2,7 +2,8 @@ package org.jetbrains.kotlinx.jupyter.messaging
 
 import org.jetbrains.kotlinx.jupyter.api.libraries.JupyterSocketType
 import org.jetbrains.kotlinx.jupyter.api.libraries.RawMessage
+import java.io.Closeable
 
-interface MessageHandler {
+interface MessageHandler : Closeable {
```
Comment (ileasile) [inline jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/messaging/MessageHandler.kt:7]: Interfaces shouldn't be closeable, implementations should be

File: jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/notebook/MutableNotebook.kt [hunk 8]
```diff
@@ -6,8 +6,9 @@ import org.jetbrains.kotlinx.jupyter.codegen.ResultsRenderersProcessor
 import org.jetbrains.kotlinx.jupyter.codegen.TextRenderersProcessorWithPreventingRecursion
 import org.jetbrains.kotlinx.jupyter.repl.EvalData
 import org.jetbrains.kotlinx.jupyter.repl.SharedReplContext
+import java.io.Closeable
 
-interface MutableNotebook : Notebook {
+interface MutableNotebook : Notebook, Closeable {
```
Comment (ileasile) [inline jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/notebook/MutableNotebook.kt:11]: And for this interface as well

File: jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/notebook/impl/NotebookImpl.kt [hunk 9]
```diff
@@ -24,20 +24,36 @@ import org.jetbrains.kotlinx.jupyter.api.libraries.Variable
...
 ) : MutableNotebook {
+    private fun `$debugMethod`() {
+        try {
+            while (true) {
+                Thread.sleep(1000)
+            }
+        } catch (_: InterruptedException) {
+        }
+    }
+    private val debugThread = thread(
```
Comment (ileasile) [inline jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/notebook/impl/NotebookImpl.kt:52]: Why should we create this object if we even don't run it? I suggest to make it nullable

File: src/main/kotlin/org/jetbrains/kotlinx/jupyter/messaging/MessageHandlerImpl.kt [hunk 10]
```diff
@@ -3,11 +3,11 @@ package org.jetbrains.kotlinx.jupyter.messaging
 import org.jetbrains.kotlinx.jupyter.api.libraries.RawMessage
 import org.jetbrains.kotlinx.jupyter.execution.JupyterExecutor
 import org.jetbrains.kotlinx.jupyter.messaging.comms.CommManagerInternal
-import org.jetbrains.kotlinx.jupyter.repl.ReplForJupyter
+import org.jetbrains.kotlinx.jupyter.repl.impl.ReplForJupyterImpl
 import java.util.concurrent.atomic.AtomicLong
 
 class MessageHandlerImpl(
-    private val repl: ReplForJupyter,
+    private val repl: ReplForJupyterImpl,
```
Comment (ileasile) [inline src/main/kotlin/org/jetbrains/kotlinx/jupyter/messaging/MessageHandlerImpl.kt:10]: Don't do this, it's a bad path. Just add an extension
```
fun Any.closeIfNeeded() {
    if (this is Closeable) close()
}
```

And use this extension to close the repl here

File: src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/impl/ReplForJupyterImpl.kt [hunk 11]
```diff
@@ -639,4 +640,8 @@ class ReplForJupyterImpl(
 
     override val host: KotlinKernelHost?
         get() = currentKernelHost
+
+    override fun close() {
+        notebook.close()
```
Comment (ileasile) [inline src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/impl/ReplForJupyterImpl.kt:645]: The same thing I've suggested for MessageHandlerImpl

**Review by ileasile — CHANGES_REQUESTED** [review]

File: jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/notebook/impl/NotebookImpl.kt (hunk 9 above)
Comment (nikolay-egorov) [reply jupyter-lib/shared-compiler/src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/notebook/impl/NotebookImpl.kt:52]: Agreed, let's do it

File: src/main/kotlin/org/jetbrains/kotlinx/jupyter/messaging/MessageHandlerImpl.kt (hunk 10 above)
Comment (nikolay-egorov) [reply src/main/kotlin/org/jetbrains/kotlinx/jupyter/messaging/MessageHandlerImpl.kt:10]: That's indeed nice move, +1

**Pull request merged**